source/fuzz/transformation_changing_memory_scope.cpp

```diff
@@ -0,0 +1,193 @@
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationChangingMemoryScope::TransformationChangingMemoryScope(
+    protobufs::TransformationChangingMemoryScope message)
+    : message_(std::move(message)) {}
+
+TransformationChangingMemoryScope::TransformationChangingMemoryScope(
+    const protobufs::InstructionDescriptor& needed_instruction,
+    uint32_t memory_scope_new_value_id) {
+  *message_.mutable_needed_instruction() = needed_instruction;
+  message_.set_memory_scope_new_value_id(memory_scope_new_value_id);
+}
+
+bool TransformationChangingMemoryScope::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // The instruction must be exist.
+  auto needed_instruction =
+      FindInstruction(message_.needed_instruction(), ir_context);
+  if (!needed_instruction) {
+    return false;
+  }
+
+  // Instruction must be atomic instruction only.
+  if (!IsAtomicInstruction(needed_instruction->opcode())) {
+    return false;
+  }
+
+  // The new value instruction must exist and must be OpConstant.
+  auto value_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_scope_new_value_id());
+  if (!value_instruction) {
+    return false;
+  }
+  if (value_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+
+  // The value instruction must be an Integer.
+  if (ir_context->get_def_use_mgr()
+          ->GetDef(value_instruction->type_id())
+          ->opcode() != SpvOpTypeInt) {
+    return false;
+  }
+
+  // The size of the integer for value instruction must be equal to 32 bits.
+  auto value_instruction_int_width = ir_context->get_def_use_mgr()
+                                         ->GetDef(value_instruction->type_id())
+                                         ->GetSingleWordInOperand(0);
+
+  if (value_instruction_int_width != 32) {
+    return false;
+  }
+
+  // The new value of memory scope must available be wider than the older.
+  auto memory_scope_in_operand =
+      GetMemoryScopeInOperandIndex(needed_instruction->opcode());
+  auto new_memory_scope_value =
+      static_cast<SpvScope>(value_instruction->GetSingleWordInOperand(0));
+  auto old_memory_scope_value = static_cast<SpvScope>(
+      ir_context->get_def_use_mgr()
+          ->GetDef(needed_instruction->GetSingleWordInOperand(
+              memory_scope_in_operand))
+          ->GetSingleWordInOperand(0));
+
+  if (!IsValidScope(new_memory_scope_value, old_memory_scope_value)) {
+    return false;
+  }
+
+  return true;
+}
+
+void TransformationChangingMemoryScope::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_instruction =
+      FindInstruction(message_.needed_instruction(), ir_context);
+
+  uint32_t needed_index =
+      GetMemoryScopeInOperandIndex(needed_instruction->opcode());
+
+  needed_instruction->SetInOperand(needed_index,
+                                   {message_.memory_scope_new_value_id()});
+}
```

Comment:
I think you need to update the def use manager and call `set_instr_block`, or just clear all analyzes.

Reply:
Done.